File: src/PythonUnitTests/IndexTricksTests.py
import unittest
import numpy as np
from nptest import nptest

class Test_test1(unittest.TestCase):

    def test_ravel_multi_index_1(self):

          return

    def test_unravel_index_1(self):

          return

    def test_mgrid_1(self):
        
         a = nptest.mgrid[0:5]
         print(a)

         print("************")

         b = nptest.mgrid[0.0:5.5]
         print(b)

         print("************")


         c = np.mgrid[0:5,0:5]
         print(c)

         print("************")

         d = np.mgrid[0:5.5,0:5.5]
         print(d)

         print("************")

         e = nptest.mgrid[3:5,4:6, 2:4.2]
         print(e)

         return;

    def test_ogrid_1(self):

         a = nptest.ogrid[0:5]
         print(a)

         print("************")

         b = nptest.ogrid[0.0:5.5]
         print(b)

         print("************")


         c = nptest.ogrid[0:5,0:5]
         print(c)

         print("************")

         d = np.ogrid[0:5.5,0:5.5]
         print(d)

         print("************")

         e = nptest.ogrid[3:5,4:6, 2:4.2]
         print(e)

         return;

    def test_c_1(self):

        a = np.c_[np.array([1,2,3]), np.array([4,5,6])]
        print(a)

        b = np.c_[np.array([[1,2,3]]), 0, 0, np.array([[4,5,6]])]
        print(b)


    def test_r_1(self):

        a = np.r_[np.array([1,2,3]), 0, 0, np.array([4,5,6])]
        print(a)

        b = np.r_[-1:1:6j, [0]*3, 5, 6]
        print(b)

        c = np.r_['0,2,0', [1,2,3], [4,5,6]]
        print(c)

        d = np.r_['1,2,0', [1,2,3], [4,5,6]]
        print(d)

        e = np.r_['r',[1,2,3], [4,5,6]]
        print(e)

    def test_s_1(self):
        return

    def test_index_exp_1(self):
        return

    def test_ix_1(self):
        return
    
 

    def test_fill_diagonal_1(self):

        a = np.zeros((3, 3), int)
        np.fill_diagonal(a, 5)
        print(a)

        a = np.zeros((3, 3, 3, 3), int)
        np.fill_diagonal(a, 4)
        print(a[0,0])
        print(a[1,1])
        print(a[2,2])

        # tall matrices no wrap
        a = np.zeros((5, 3),int)
        np.fill_diagonal(a, 4)
        print(a)

        # tall matrices wrap
        a = np.zeros((5, 3),int)
        np.fill_diagonal(a, 4, wrap = True)
        print(a)

        # wide matrices wrap
        a = np.zeros((3, 5),int)
        np.fill_diagonal(a, 4, wrap = True)
        print(a)

        return

    def test_diag_indices_1(self):

        di = np.diag_indices(4)
        print(di)

        a = np.arange(16).reshape(4, 4)
        a[di] = 100
        print(a)

        return
    
    def test_diag_indices_from_1(self):

        a = np.arange(16).reshape(4, 4)
        di = np.diag_indices_from(a)
        print(di)
        return


if __name__ == '__main__':
    unittest.main()

File: src/NumpyDotNet/UnitTests/NumpyDotNetTests/IndexTricksTests.cs
using System;
using System.Numerics;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using NumpyDotNet;
using System.Collections.Generic;
using System.Text;
using System.Linq;
using NumpyLib;

namespace NumpyDotNetTests
{
    [TestClass]
    public class IndexTricksTests : TestBaseClass
    {
#if NOT_PLANNING_TODO
        [Ignore] // not implemented yet.  Too much work
        [TestMethod]
        public void xxx_test_ravel_multi_index_1()
        {

        }

        [Ignore] // not implemented yet.  Too much work
        [TestMethod]
        public void xxx_test_unravel_index_1()
        {

        }

        [Ignore] // not implemented yet.  Too much work
        [TestMethod]
        public void xxx_test_c_1()
        {

        }

        [Ignore] // not implemented yet.  Too much work
        [TestMethod]
        public void xxx_test_r_1()
        {

        }

        [Ignore] // not implemented yet.  Too much work
        [TestMethod]
        public void xxx_test_s_1()
        {

        }

        [Ignore] // not implemented yet.  Too much work
        [TestMethod]
        public void xxx_test_index_exp_1()
        {

        }

        [Ignore] // not implemented yet.  Too much work
        [TestMethod]
        public void xxx_test_ix_1()
        {

        }

#endif


        [TestMethod]
        public void test_mgrid_1()
        {
            var a = (ndarray)np.mgrid(new Slice[] { new Slice(0, 5) });
            print(a);
            AssertArray(a, new Int32[] { 0, 1, 2, 3, 4 });
            print("************");

            var b = (ndarray)np.mgrid(new Slice[] { new Slice(0.0, 5.5) });
            print(b);
            AssertArray(b, new double[] { 0.0, 1.0, 2.0, 3.0, 4.0, 5.0 });
            print("************");

            var c = (ndarray)np.mgrid(new Slice[] { new Slice(0, 5), new Slice(0, 5) });
            print(c);

            var ExpectedCArray = new Int32[,,]
                {{{0, 0, 0, 0, 0},  {1, 1, 1, 1, 1},  {2, 2, 2, 2, 2},  {3, 3, 3, 3, 3},  {4, 4, 4, 4, 4}},
                 {{0, 1, 2, 3, 4},  {0, 1, 2, 3, 4},  {0, 1, 2, 3, 4},  {0, 1, 2, 3, 4},  {0, 1, 2, 3, 4}}};
            AssertArray(c, ExpectedCArray);


            print("************");

            var d = (ndarray)np.mgrid(new Slice[] { new Slice(0, 5.5), new Slice(0, 5.5) });
            print(d);
            var ExpectedDArray = new double[,,]
                {{{0, 0, 0, 0, 0, 0},  {1, 1, 1, 1, 1, 1},  {2, 2, 2, 2, 2, 2},  {3, 3, 3, 3, 3, 3},  {4, 4, 4, 4, 4, 4}, {5, 5, 5, 5, 5, 5}},
                 {{0, 1, 2, 3, 4, 5},  {0, 1, 2, 3, 4, 5},  {0, 1, 2, 3, 4, 5},  {0, 1, 2, 3, 4, 5},  {0, 1, 2, 3, 4, 5}, {0, 1, 2, 3, 4, 5}}};
            AssertArray(d, ExpectedDArray);

            print("************");

            var e = (ndarray)np.mgrid(new Slice[] { new Slice(3, 5), new Slice(4, 6), new Slice(2, 4.2) });
            print(e);
            var ExpectedEArray = new double[,,,]
                {
                    {{{3, 3, 3}, {3, 3, 3}}, {{4, 4, 4}, {4, 4, 4}}},
                    {{{4, 4, 4}, {5, 5, 5}}, {{4, 4, 4}, {5, 5, 5}}},
                    {{{2, 3, 4}, {2, 3, 4}}, {{2, 3, 4}, {2, 3, 4}}},
                };
            AssertArray(e, ExpectedEArray);

        }


        [TestMethod]
        public void test_ogrid_1()
        {
            var a = (ndarray)np.ogrid(new Slice[] { new Slice(0, 5) });
            print(a);
            AssertArray(a, new Int32[] { 0, 1, 2, 3, 4 });
            print("************");

            var b = (ndarray)np.ogrid(new Slice[] { new Slice(0.0, 5.5) });
            print(b);
            AssertArray(b, new double[] { 0.0, 1.0, 2.0, 3.0, 4.0, 5.0 });
            print("************");

            var c = (ndarray[])np.ogrid(new Slice[] { new Slice(0, 5), new Slice(0, 5) });
            print(c);
            AssertArray(c[0], new Int32[,] { { 0 }, { 1 }, { 2 }, { 3 }, { 4 } });
            AssertArray(c[1], new Int32[,] { { 0, 1, 2, 3, 4 } });


            print("************");

            var d = (ndarray[])np.ogrid(new Slice[] { new Slice(0, 5.5), new Slice(0, 5.5) });
            print(d);
            AssertArray(d[0], new Int32[,] { { 0 }, { 1 }, { 2 }, { 3 }, { 4 }, { 5 } });
            AssertArray(d[1], new Int32[,] { { 0, 1, 2, 3, 4, 5 } });

            print("************");

            var e = (ndarray[])np.ogrid(new Slice[] { new Slice(3, 5), new Slice(4, 6), new Slice(2, 4.2) });
            print(e);
            AssertArray(e[0], new Int32[,,] { { { 3 } }, { { 4 } } });
            AssertArray(e[1], new Int32[,,] { { { 4 }, { 5 } } });
            AssertArray(e[2], new Int32[,,] { { { 2, 3, 4 } } });

        }

        [TestMethod]
        public void test_fill_diagonal_1()
        {
            var a = np.zeros((3, 3), np.Int32);
            np.fill_diagonal(a, 5);
            AssertArray(a, new int[,] { { 5, 0, 0 }, { 0, 5, 0 }, { 0, 0, 5 } });
            print(a);

            a = np.zeros((3, 3, 3, 3), np.Int32);
            np.fill_diagonal(a, 4);
            AssertArray(a[0, 0] as ndarray, new int[,] { { 4, 0, 0 }, { 0, 0, 0 }, { 0, 0, 0 } });
            print(a[0, 0]);
            AssertArray(a[1, 1] as ndarray, new int[,] { { 0, 0, 0 }, { 0, 4, 0 }, { 0, 0, 0 } });
            print(a[1, 1]);
            AssertArray(a[2, 2] as ndarray, new int[,] { { 0, 0, 0 }, { 0, 0, 0 }, { 0, 0, 4 } });
            print(a[2, 2]);

            // tall matrices no wrap
            a = np.zeros((5, 3), np.Int32);
            np.fill_diagonal(a, 4);
            AssertArray(a, new int[,] { { 4, 0, 0 }, { 0, 4, 0 }, { 0, 0, 4 }, { 0, 0, 0 }, { 0, 0, 0 } });
            print(a);

            // tall matrices wrap
            a = np.zeros((5, 3), np.Int32);
            np.fill_diagonal(a, 4, wrap:true);
            AssertArray(a, new int[,] { { 4, 0, 0 }, { 0, 4, 0 }, { 0, 0, 4 }, { 0, 0, 0 }, { 4, 0, 0 } });
            print(a);

            // wide matrices wrap
            a = np.zeros((3, 5), np.Int32);
            np.fill_diagonal(a, 4, wrap: true);
            AssertArray(a, new int[,] { { 4, 0, 0, 0, 0 }, { 0, 4, 0, 0, 0 }, { 0, 0, 4, 0, 0 } });
            print(a);


        }

        [TestMethod]
        public void test_diag_indices_1()
        {
            var di = np.diag_indices(4);
            AssertArray(di[0], new Int32[] { 0, 1, 2, 3 });
            AssertArray(di[1], new Int32[] { 0, 1, 2, 3 });
            print(di);

            var a = np.arange(16).reshape((4, 4));
            a[di] = 100;

            AssertArray(a, new Int32[,] { {100,1,2,3 }, {4,100,6,7 }, {8, 9, 100, 11 }, { 12,13,14,100 } });
            print(a);

            return;

        }

        [TestMethod]
        public void test_diag_indices_from_1()
        {
            var a = np.arange(16).reshape((4, 4));
            var di = np.diag_indices_from(a);
            AssertArray(di[0], new Int32[] { 0, 1, 2, 3 });
            AssertArray(di[1], new Int32[] { 0, 1, 2, 3 });
            print(di);
        }


    }
}
